pyphare/pyphare_tests/test_pharesee/test_geometry.py

```diff
@@ -0,0 +1,528 @@
+
+
+
+
+
+    def test_level_ghost_boxes(self):
+
+        expected = {
+
+            1: [
+                {"boxes": [Box(9, 9),]},
+                {"boxes": [Box(60, 60),] },
+                {"boxes": [Box(63, 63),]},
+                {"boxes": [Box(112, 112),]}
+            ]
+        }
+
+        lvl_gaboxes = level_ghost_boxes(self.hierarchy)
+        for ilvl  in range(1, len(self.hierarchy.patch_levels)):
+            for actual, exp in zip(lvl_gaboxes[ilvl], expected[ilvl]):
+
+                act_boxes = actual["boxes"]
+                exp_boxes = exp["boxes"]
+
+                for act_box, exp_box in zip(act_boxes, exp_boxes):
+                    self.assertEqual(act_box, exp_box)
+
+
+
+
+
+    def test_overlapped_fields_are_equal(self):
+
+        overlaps = hierarchy_overlaps(self.hierarchy)
+
+        for ilvl, lvl in enumerate(self.hierarchy.patch_levels):
+
+            for overlap in overlaps[ilvl]:
+
+                pd1, pd2 = overlap["pdatas"]
+                box      = overlap["box"]
+                offsets  = overlap["offset"]
+
+                self.assertEqual(pd1.quantity, pd2.quantity)
+
+                if pd1.quantity == 'field':
+
+                    # we need to transform the AMR overlap box, which is thus
+                    # (because AMR) common to both pd1 and pd2 into local index
+                    # boxes that will allow to slice the data
+
+                    # the patchData ghost box that serves as a reference box
+                    # to transfrom AMR to local indexes first needs to be
+                    # shifted by the overlap offset associated to it
+                    # this is because the overlap box has been calculated from
+                    # the intersection of possibly shifted patch data ghost boxes
+
+                    loc_b1 = boxm.amr_to_local(box, boxm.shift(pd1.ghost_box, offsets[0]))
+                    loc_b2 = boxm.amr_to_local(box, boxm.shift(pd2.ghost_box, offsets[1]))
+
+                    data1 = pd1.dataset
+                    data2 = pd2.dataset
+
+                    slice1 = data1[loc_b1.lower:loc_b1.upper+1]
+                    slice2 = data2[loc_b2.lower:loc_b2.upper+ 1]
+
+                    self.assertTrue(np.allclose(slice1, slice2, atol=1e-12))
+
+
+
+
+
+
+
+    def test_overlapped_particledatas_have_identical_particles(self):
+
+        from copy import copy
+        overlaps = hierarchy_overlaps(self.hierarchy)
+
+        for ilvl, lvl in enumerate(self.hierarchy.patch_levels):
+
+            for overlap in overlaps[ilvl]:
+
+                pd1, pd2 = overlap["pdatas"]
+                box      = overlap["box"]
+                offsets  = overlap["offset"]
+
+                self.assertEqual(pd1.quantity, pd2.quantity)
+
+                if pd1.quantity == "particles":
+
+                    # the following uses 'offset', we need to remember that offset
+                    # is the quantity by which a patch has been moved to detect
```

Comment:
sorry, I do not get this one...

Reply:
the if ? 
the test is named `test_overlapped_particledatas_have_identical_particles`
so the if is here to only select particledatas among patchdatas